Wheel plowing approximation in ODEPhysics

scpeters opened the pull request:
As noted in https://github.com/osrf/gazebo/issues/3085, the distribution of contact forces can be quite different for a wheel driving on rigid terrain and deformable terrain.  Contact for rigid wheel on rigid terrain can be modeled with a single contact point and normal direction perpendicular to the surface. For a rigid wheel on deformable terrain, the contact occurs over a larger area. The contact pressure distribution can be approximated as an equivalent point force at the centroid of the pressure distribution. At rest, a wheel on a deformable surface may sink into the surface with a symmetric distribution of normal contact pressure, such that the equivalent normal force vector is parallel to the undisturbed surface normal, which is vertical in the illustration below.

![wheel_plowing](https://user-images.githubusercontent.com/3650755/150075133-b786194f-57d3-4264-92fe-b469feb3078c.png)

While rolling, the sinkage of the wheel may cause a plowing effect, in which the terrain material is bunched up in the direction of travel and compacted in its wake. This can cause a significant forward bias in the distribution of normal contact pressure and an equivalent normal force vector that is inclined relative to the undisturbed surface normal. In the illustration above, the equivalent normal force opposes the motion of the wheel, requiring tractive torque to be applied in order to maintain a steady velocity. Even when flat, the plowing effect means that driving on a deformable surface is energetically equivalent to driving uphill on a rigid surface.

This pull request implements an approximation of longitudinal wheel plowing effects by displacing contact points and inclining the contact normals to replicate the behavior illustrated in the figure above. This is implemented in `ODEPhysics::Collide`, which is a callback function called for each pair of colliding bodies. To enable the plowing effect, special SDFormat tags must be added to the terrain (`<gz:plowing_terrain />`) and the wheels (`<gz:plowing_wheel />`). The wheel must specify `<fdir1>` in its `<surface>`, and the `<gz:plowing_wheel>` tag must also contain `<max_degrees>` and `<saturation_velocity>` and may optionally contain `<deadband_velocity>`. See the `plowing_effect_*` models in `test/models` for examples.

scpeters committed:
ODEPhysics: approximate wheel plowing effects

This builds on the implementation of the plowing effect
from the aditya/contact_pts branch to implement a model
of longitudinal wheel plowing effects on deformable
terrain discussed in #3085.

The approach rotates contact points and normals forward
to approximate the normal pressure distribution when
plwing through deformable soil.

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Collision::WorldLinearVel: account for pose offset

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

adityapande-1995 commented:
Left : Vehicle using the plowing effect, right : without the plowing effect
![image](https://user-images.githubusercontent.com/18661155/151893580-54cd3150-c966-4d86-aad0-ec06e0847018.png)

Wheels sinking into the ground : 
![image](https://user-images.githubusercontent.com/18661155/151895124-b49ce267-a16c-4a63-9132-183d2281b5d8.png)

adityapande-1995 committed:
Updates by Aditya

* Removed unused code
* Added example world and model directory
* Added plowing_wheel and plowing_terrain sdf tags

Signed-off-by: Aditya <aditya050995@gmail.com>

adityapande-1995 commented:
To run the example : ``gazebo --verbose worlds/plowing_effect_demo.world``

scpeters committed:
Refactor helper function and SDFormat tags

* Rename GetRotationAngle as deadband_saturation
* Use gz instead of ignition for the xmlns
* Nest wheel parameters inside <gz:plowing_wheel />
* Move plowing_effect models to test/models
* Comment out ROS plugins from demo world

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters commented:
I've updated this and marked it ready for review

WilliamLewww reviewed (approved):
A minor warning:
```
[Err] [InsertModelWidget.cc:403] Missing model.config for model "/gazebo-ws/shared/gazebo/test/models/test_nested_urdf"
[Err] [InsertModelWidget.cc:403] Missing model.config for model "/gazebo-ws/shared/gazebo/test/models/testdb"
```

but looks good to me!

scpeters commented on test/models/plowing_effect_ground_plane/model.config:
```diff
@@ -0,0 +1,16 @@
+
+  <author>
+    <name>Nate Koenig</name>
+    <email>nate@osrfoundation.org</email>
```
@adityapande-1995 I'll add you as a co-author on these new models

scpeters replied:
https://github.com/osrf/gazebo/pull/3164/commits/cd9a5869193cbf59ba5e4b311856aa20668e0a20

scpeters commented on gazebo/physics/ode/ODEPhysics.cc:
```diff
@@ -1692,3 +1800,41 @@ bool ODEPhysics::GetParam(const std::string &_key, boost::any &_value) const
+
+/// \brief A mathematical function that represents a saturation function
+/// that increases linearly until reaching a constant value with the addition
+/// of a deadband near zero. THe shape of this function for positive input
```
capitalization: `THe` -> `The`

scpeters replied:
https://github.com/osrf/gazebo/pull/3164/commits/cd9a5869193cbf59ba5e4b311856aa20668e0a20

scpeters commented on gazebo/physics/ode/ODEPhysics.cc:
```diff
@@ -1692,3 +1800,41 @@ bool ODEPhysics::GetParam(const std::string &_key, boost::any &_value) const
+{
+  if (std::abs(_input) <= std::abs(_deadband))
+  {
+    return 0;
```
return 0.0

scpeters replied:
https://github.com/osrf/gazebo/pull/3164/commits/cd9a5869193cbf59ba5e4b311856aa20668e0a20

scpeters commented on gazebo/physics/ode/ODEPhysics.cc:
```diff
@@ -1161,6 +1164,111 @@ void ODEPhysics::Collide(ODECollision *_collision1, ODECollision *_collision2,
+
+      // compute slope
+      double slope = plowingMaxAngle.Radian() /
+          (plowingSaturationVelocity - plowingDeadbandVelocity);
```
check that saturation velocity is larger than deadband velocity

scpeters replied:
this ended up being a bigger change, as I moved the SDF parameter parsing to a static helper function in ODECollision and added parameter checks there

https://github.com/osrf/gazebo/pull/3164/commits/dd64a81cc396cc1ef7adf27fd167867bb0bc1570

scpeters committed:
Minor fixes

* Add Aditya as model author
* Return 0.0 instead of 0
* Fix typo: THe

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Parse wheel plowing parameters in helper function

Add static helper function to ODECollision for parsing wheel
parsing parameters. Use _verbose flag to optionally print
error messages if invalid parameter values are detected.

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

deepanshubansal01 committed:
plowing effect unit test (#3229)

This PR implements unit tests for plowing effect feature.

Signed-off-by: deepanshu <deepanshubansal01@gmail.com>

scpeters committed:
Merge branch 'gazebo11' into scpeters/plowing

scpeters committed:
Minor test refactoring

* Remove unused variable callbackCountRigidTerrain_
* Fix whitespace in model.sdf
* plowing_effect_test.world: set model pose instead
  of link pose, and raise spheres above ground level
  by setting initial Z coordinate to value of radius

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Update test: tricycle on deformable terrain

* Fix calculation of longitudinal unit vector and
  expect that longitudinal velocity and contact unit
  normal vector oppose each other.
* Update model plowing parameters to ensure max plowing
  is reached.
* Test contacts at rear wheel instead of front since
  they are more stable and expect matching contact
  point at each time step.
* Print one debug message for each plowing state.

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
plowing_effect_test.world: update contact params

Improve stability of contact points through softened
kp and kd, and small max_vel and min_depth values.

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters commented:
tests have been kindly provided by @deepanshubansal01 in https://github.com/osrf/gazebo/pull/3229, and I have made some small adjustments:

* https://github.com/osrf/gazebo/pull/3164/commits/b6cf816760253292a4832bd29d3515e2563311e6: minor refactoring of test code and models
* https://github.com/osrf/gazebo/pull/3164/commits/ecedcc198c2bb26444928a7691018d0c632b82bd: adjustments to tricycle on deformable terrain test: update parameters to ensure all plowing states are reached, fix one of the calculations, and improve console messaging
* https://github.com/osrf/gazebo/pull/3164/commits/5432fe58a916aeefa6b2b82116bb863b63f6f320: improve contact parameters in plowing_effect_test.world to improve contact point stability

adityapande-1995 reviewed (approved):
Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of ``fdir1`` for now.

deepanshubansal01 reviewed (approved):
Looks good to me as well.

scpeters commented:
> Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.

To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction.  I'll go ahead and add some notes about this in the example model / world

scpeters committed:
Merge branch 'gazebo11' into scpeters/plowing

scpeters committed:
Improve description of plowing tricycle model

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Move, rename test worlds

* Move tricycle demo world to test folder since its models
  aren't installed and rename to plowing_effect_tricycle_demo.world
* Rename spheres test world to plowing_effect_spheres_test.world

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Add documentation to demo world

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters committed:
Add documentation to spheres test world

Signed-off-by: Steve Peters <scpeters@openrobotics.org>

scpeters commented:
> > Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.
> 
> To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction. I'll go ahead and add some notes about this in the example model / world

I've added documentation in several recent commits; let me know what you think @adityapande-1995 and @deepanshubansal01

adityapande-1995 commented:
> > > Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.
> > 
> > 
> > To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction. I'll go ahead and add some notes about this in the example model / world
> 
> I've added documentation in several recent commits; let me know what you think @adityapande-1995 and @deepanshubansal01

Looks good to me !

deepanshubansal01 commented:
> > > Tests look good ! I feel we should document in the examples / code that the plowing effect only applies in the direction of `fdir1` for now.
> > 
> > 
> > To be precise, the plowing effect applies in the direction normal to `fdir1` and the surface normal. If `fdir1` is the wheel's lateral direction, then the plowing effect will apply in the wheel's longitudinal direction. I'll go ahead and add some notes about this in the example model / world
> 
> I've added documentation in several recent commits; let me know what you think @adityapande-1995 and @deepanshubansal01

Looks good to me as well !
